stdlib/asm/crypto/hashes/rpo.masm

```diff
@@ -100,3 +100,161 @@ export.hash_memory
   # drop start_addr/end_addr (4 cycles)
   movup.4 drop movup.4 drop
 end
+
+#! Computes hash of Felt values starting at the specified memory address.
+#!
+#! This procedure divides the hashing process into two parts: hashing pairs of words using 
+#! `absorb_double_words_from_memory` procedure and hashing the remaining values using the `hperm`
+#! instruction. 
+#!
+#! Inputs:  [ptr, num_elements]
+#! Outputs: [HASH]
+#! Cycles: 
+#! - If number of elements divides by 8: 47 cycles + 3 * words
+#! - Else: 180 cycles + 3 * words
+#!
+#! Panics if number of inputs equals 0.
+export.hash_memory
+    # move number of inputs to the top of the stack 
+    swap
+    # => [num_elements, ptr]
+
+    # check that number of inputs greater than 0
+    dup eq.0 assertz
+    # => [num_elements, ptr]
+
+    # get the number of double words
+    u32divmod.8 swap
+    # => [num_elements/8, num_elements%8, ptr]
+
+    # get the end_addr for hash_memory_even procedure (end address for pairs of words)
+    mul.2 dup.2 add movup.2
+    # => [ptr, end_addr, num_elements%8]
+
+    # get the capacity element which is equal to num_elements%8
+    dup.2
+    # => [capacity, ptr, end_addr, num_elements%8]
+    
+    # prepare hasher state for RPO permutation 
+    push.0.0.0 padw padw 
+    # => [C, B, A, ptr, end_addr, num_elements%8]
+
+    # hash every pair of words
+    exec.absorb_double_words_from_memory
+    # => [C', B', A', ptr', end_addr, num_elements%8] where ptr' = end_addr
+
+    # hash remaining input values if there are any left
+    # if num_elements%8 is ZERO and there are no elements to hash
+    dup.14 eq.0
+    if.true
+      # clean the stack
+      exec.squeeze_digest
+      swapw dropw
```

Comment:
I'm now sure I understand why `swapw dropw` works here. Isn't the state of the stack before these instructions look like so:
```
[B', ptr', end_addr, num_elements%8]
```
And so there are only 3 extra elements after `B'`. This means we drop one extra element from the stack - no?

Let's make sure we test for this in the tests.

Reply:
Ah, I see. I unconsciously assumed that there are no other elements on the stack, so everything except current values are `ZERO`'s so we can safely drop them. But in real life use cases there would be other elements, now I see what did you mean. I will fix this, sorry.